### PR TITLE
[Polkadot]: Runtime network id for polkadot

### PR DESCRIPTION
**Pull request by Milerius**
## Description 

- [x] Introduce networkId
- [x] C Interface to query networkID
- [x] Update CoinInfoData codegen template
- [x] AnyAddress unit tests with Acala network
- [x] Address validation unit tests with Acala network
- [x] Being able to create an AnyAddress with a network ID
- [x] Change ValidateAddress to take a variant, make it more generic, avoid adding more field types in the future

## Notes

- Keep Kusama blockchain type to avoid breaking change, next Polkadot chain addition can use Polkadot as blockchain type

## How to test

Run C++ unit tests

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: src/proto/Polkadot.proto [hunk 1]
```diff
@@ -4,10 +4,10 @@ package TW.Polkadot.Proto;
 option java_package = "wallet.core.jni.proto";
 
 // Known networks
-enum Network {
+/*enum Network {
     POLKADOT = 0;
     KUSAMA = 2;
-}
+}*/
```
Comment (Milerius) [inline src/proto/Polkadot.proto:10]: to remove later

File: src/proto/Polkadot.proto [hunk 2]
```diff
@@ -142,7 +142,7 @@ message SigningInput {
     bytes private_key = 8;
 
     // Network type
-    Network network = 9;
+    uint32 network = 9;
```
Comment (Milerius) [inline src/proto/Polkadot.proto:145]: this is a breaking change for end user app, but required for future runtime support

File: src/Bitcoin/Entry.cpp [hunk 3]
```diff
@@ -23,20 +22,29 @@ bool Entry::validateAddress(TWCoinType coin, const std::string& address, byte p2
     case TWCoinTypeQtum:
     case TWCoinTypeViacoin:
     case TWCoinTypeBitcoinGold:
-        return SegwitAddress::isValid(address, hrp) || Address::isValid(address, {{p2pkh}, {p2sh}});
-
+        if (auto* prefix = std::get_if<UTXOPrefix>(&addressPrefix); prefix) {
```
Comment (catenocrypt) [inline src/Bitcoin/Entry.cpp:25]: I think `UTXOPrefix` is not the best name, it is an address type / address prefix, not specific to UTXOs.

Comment (Milerius) [reply src/Bitcoin/Entry.cpp:25]: > I think `UTXOPrefix` is not the best name, it is an address type / address prefix, not specific to UTXOs.

Trying to find an appropriate name here, I've checked the registry.json and only UTXO coins use p2sh/p2pkh. Which name we can use instead?

File: src/CoinEntry.h [hunk 4]
```diff
@@ -17,22 +17,36 @@
...
 namespace TW {
 
 typedef std::vector<std::pair<Data, Data>> HashPubkeyList;
 
+struct UTXOPrefix {
+    TW::byte p2pkh;
+    TW::byte p2sh;
+};
+
+using HRPPrefix = const char *;
```
Comment (catenocrypt) [inline src/CoinEntry.h:32]: Maybe we could call this `Bech32Prefix`, a more specific name.

File: src/CoinEntry.h [hunk 5]
```diff
@@ -17,22 +17,36 @@
...
 #include <string>
 #include <vector>
+#include <variant>
 #include <utility>
 
 namespace TW {
 
 typedef std::vector<std::pair<Data, Data>> HashPubkeyList;
 
+struct UTXOPrefix {
```
Comment (catenocrypt) [inline src/CoinEntry.h:27]: Need to find a better name here... This is Bitcoin (+fork) specific, for 'legacy' non-segwit, Base58-encoded addresses.  `Base58Prefix`?

File: registry.json [hunk 6]
```diff
@@ -1022,6 +1022,7 @@
     "curve": "ed25519",
     "publicKeyType": "ed25519",
     "addressHasher": "keccak256",
+    "networkPrefix": 0,
```
Comment (hewigovens) [inline registry.json:1025]: let's use `ss58Prefix`? https://github.com/paritytech/ss58-registry/

File: src/Decred/Entry.h [hunk 7]
```diff
@@ -14,7 +14,7 @@ namespace TW::Decred {
 /// Note: do not put the implementation here (no matter how simple), to avoid having coin-specific includes in this file
 class Entry final : public CoinEntry {
 public:
-     bool validateAddress(TWCoinType coin, const std::string& address, TW::byte p2pkh, TW::byte p2sh, const char* hrp) const;
+    bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```
Comment (hewigovens) [inline src/Decred/Entry.h:17]:
```suggestion
     bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```

File: src/EOS/Entry.h [hunk 8]
```diff
@@ -14,7 +14,7 @@ namespace TW::EOS {
 /// Note: do not put the implementation here (no matter how simple), to avoid having coin-specific includes in this file
 class Entry final : public CoinEntry {
 public:
-     bool validateAddress(TWCoinType coin, const std::string& address, TW::byte p2pkh, TW::byte p2sh, const char* hrp) const;
+    bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```
Comment (hewigovens) [inline src/EOS/Entry.h:17]:
```suggestion
     bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```

File: src/Elrond/Entry.h [hunk 9]
```diff
@@ -14,7 +14,7 @@ namespace TW::Elrond {
 /// Note: do not put the implementation here (no matter how simple), to avoid having coin-specific includes in this file
 class Entry final: public CoinEntry {
 public:
-     bool validateAddress(TWCoinType coin, const std::string& address, TW::byte p2pkh, TW::byte p2sh, const char* hrp) const;
+    bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```
Comment (hewigovens) [inline src/Elrond/Entry.h:17]:
```suggestion
     bool validateAddress(TWCoinType coin, const std::string& address, const PrefixVariant& addressPrefix) const;
```

Comment (hewigovens) [reply src/Elrond/Entry.h:17]: There are a few other places, maybe we can enable auto format now

File: tests/chains/Polkadot/SignerTests.cpp [hunk 10]
```diff
@@ -43,7 +44,7 @@ TEST(PolkadotSigner, SignTransfer_9fd062) {
         EXPECT_EQ(address.string(), addressThrow2);
     }
     input.set_private_key(privateKeyThrow2.bytes.data(), privateKeyThrow2.bytes.size());
-    input.set_network(Proto::Network::POLKADOT);
+    input.set_network(ss58Prefix(TWCoinType::TWCoinTypePolkadot));
```
Comment (hewigovens) [inline tests/chains/Polkadot/SignerTests.cpp:47]: not need to use `TWCoinType::`?

**Review by hewigovens — COMMENTED** [review]
Don't forget changing Entry.erb

File: src/Elrond/Entry.h (hunk 9 above)
Comment (Milerius) [reply src/Elrond/Entry.h:17]: Auto-format after files are already pushed doesn't work in my IDE, but I fixed the above place^

**Review by catenocrypt — DISMISSED** [review]

File: src/proto/Polkadot.proto (hunk 2 above)
Comment (hewigovens) [reply src/proto/Polkadot.proto:145]: For breaking changes, need to mention it in next release notes (how to use new code)

File: src/proto/Polkadot.proto (hunk 1 above)
Comment (hewigovens) [reply src/proto/Polkadot.proto:10]: or we can keep it for compatibility? Then cast enum raw value to `uint32`

**Review by hewigovens — COMMENTED** [review]
left some comments

File: src/proto/Polkadot.proto (hunk 1 above)
Comment (Milerius) [reply src/proto/Polkadot.proto:10]: I think it's better to remove it and mention it to the release, it's an easy upgrade.

**Review by miloserdow — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Pull request merged**